### ISSUE TITLE
(aws-batch): job-role attribute not populated

### ISSUE THREAD

**Issue by npvisual** [aws/aws-cdk#16020]
The [JobDefinitionContainer](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinitionContainer.html#jobdefinitioncontainer) in the AWS Batch service is, from what I understand, supposed to hold a [job_role](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinitionContainer.html#aws_cdk.aws_batch.JobDefinitionContainer.job_role) that gets created automatically if none is provided during instantiation. 

>  The IAM role that the container can assume for AWS permissions. Default: - An IAM role will created.

However when requesting this role to assign the proper grants on a database key, that attribute is undefined.

### Reproduction Steps

The `JobDefinitionContainer` is created using the following code and coupled with the traditional AWS Batch components : job definition, job queue and compute environment.

```python
        # Container definition (createdb)
        batch_createdb_container = batch.JobDefinitionContainer(
            image=ecs.ContainerImage.from_ecr_repository(repository=repo, tag=app_tag),
            command=["-m", "app.scripts.createdb"],
            environment=environment_variables,
            memory_limit_mib=2048,
        )
```

The decrypt permission (grant) is made with the following : 

```python
        database.secret.encryption_key.grant_decrypt(batch_createdb_container.job_role)
```

Note that all the batch components are created if that grant above is commented out.

### What did you expect to happen?

I would expect that the `job_role` be populated properly (as described in the documentation) so that it can be used in grant statements as depicted above.

### What actually happened?

The following error is received when a `cdk diff` or `cdk deploy` is performed : 

```
....
  File "/path/furley-0NV7C4_t-py3.9/lib/python3.9/site-packages/aws_cdk/aws_kms/__init__.py", line 1515, in grant_decrypt
    return typing.cast(aws_cdk.aws_iam.Grant, jsii.invoke(self, "grantDecrypt", [grantee]))
  File "/path/furley-0NV7C4_t-py3.9/lib/python3.9/site-packages/jsii/_kernel/__init__.py", line 128, in wrapped
    return _recursize_dereference(kernel, fn(kernel, *args, **kwargs))
  File "/path/furley-0NV7C4_t-py3.9/lib/python3.9/site-packages/jsii/_kernel/__init__.py", line 340, in invoke
    response = self.provider.invoke(
  File "/path/furley-0NV7C4_t-py3.9/lib/python3.9/site-packages/jsii/_kernel/providers/process.py", line 359, in invoke
    return self._process.send(request, InvokeResponse)
  File "/path/furley-0NV7C4_t-py3.9/lib/python3.9/site-packages/jsii/_kernel/providers/process.py", line 326, in send
    raise JSIIError(resp.error) from JavaScriptError(resp.stack)
jsii.errors.JSIIError: Got 'undefined' for non-optional instance of {"name":"grantee","type":{"fqn":"@aws-cdk/aws-iam.IGrantable"}}
Subprocess exited with error 1
```

### Environment

  - **CDK CLI Version  :** 1.117.0 (build 0047c98)
  - **Framework Version:** 1.117.0
  - **Node.js Version:** v15.8.0
  - **OS               :** MacOS v11.4 (20F71)
  - **Language (Version):** Python 3.9.5

### Other

I am essentially trying to do the equivalent of the following code (Lambda) but for Batch : 

```python
       database.secret.encryption_key.grant_decrypt(lambda_container.grant_principal)
       database.secret.grant_read(lambda_container.grant_principal)
```
I just can't seem to find a valid `IGrantable` in any of the Batch constructs.
If there's a better way to assign a read grant for a database secret to a batch job, please let me know !

--- 

This is :bug: Bug Report

**Comment by npvisual** [comment]
Hi ! Any updates on this issue ? Anything I can do to help ?

Would there be any workaround ? I was trying to figure out a way to retrieve the role that gets automatically assigned to the ComputeResources construct, or to set a role I created for it. Unfortunately it only takes a `str` as input or only gives a `str` as property value.

Any help would be greatly appreciated as I am currently stuck having to set those permissions manually every time I deploy.

Thanks !

**Comment by npvisual** [comment]
Quick update : 

I realized I was initially asking about the [job_role](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinitionContainer.html#aws_cdk.aws_batch.JobDefinitionContainer.job_role) for the  JobDefinitionContainer construct.

However it appears that the [instance_role](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/ComputeResources.html#aws_cdk.aws_batch.ComputeResources.instance_role) of the [ComputeResources](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/ComputeResources.html) is the role that is being assumed by the instances when running a batch job.

Should I update the title of the issue ?

**Comment by npvisual** [comment]
So after a little digging, I can work around the lack of IRole / IGrantable in the Batch APIs, by using the following (Python code) : 

```python
        # ECS instance role from the compute environment
        # We have to cheat our way out of :
        # https://github.com/aws/aws-cdk/issues/16020
        ecs_instance_role: iam.Role = batch_compute_environment.node.find_child(
            "Ecs-Instance-Role"
        )
```

and then we can apply the `grant_read` and `grant_decrypt` convenience methods : 

```python
        # We give the batch compute resource role the proper permission
        # to read the DB cluster's secret (and therefore a `decrypt` permission
        # on the secret's key)
        database.secret.grant_read(ecs_instance_role)
        database.secret.encryption_key.grant_decrypt(ecs_instance_role)
```

I have verified this adds an inline policy to the "Ecs_instance_role" that gets assumed by the service for the secret and the key.

While this works, it would be much better to have something like what's offered by the Lambda APIs (see initial description above).

**Comment by madeline-k** [comment]
Hey @npvisual, can you please provide the code for your stack containing the workaround? Knowing how you've set things up will help me understand the issue better.

**Comment by madeline-k** [comment]
The [JobDefinitionContainer](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinitionContainer.html#jobdefinitioncontainer) is not ever modified by the CDK it is just an object to hold properties for when you create a [JobDefinition](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinition.html). So if a role is created for you, this will be created somewhere else, and might be accessible somewhere else. but not in [job_role](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinitionContainer.html#aws_cdk.aws_batch.JobDefinitionContainer.job_role)

**Comment by npvisual** [comment]
> Hey @npvisual, can you please provide the code for your stack containing the workaround? Knowing how you've set things up will help me understand the issue better.

@madeline-k , sure. Gonna need to write separate code to extract it from our application code. Should be able to send that tomorrow.

**Comment by npvisual** [comment]
> The [JobDefinitionContainer](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinitionContainer.html#jobdefinitioncontainer) is not ever modified by the CDK it is just an object to hold properties for when you create a [JobDefinition](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinition.html). So if a role is created for you, this will be created somewhere else, and might be accessible somewhere else. but not in [job_role](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinitionContainer.html#aws_cdk.aws_batch.JobDefinitionContainer.job_role)

Not sure I understand your explanation. The documentation clearly states that, for both constructs `JobDefinitionContainer` and `ComputeResources` , they each have an attribute, `job_role` and `instance_role` respectively, that can be queried.

There are at least 2 roles that get created automatically for you when the batch stack is created : 
* one role assumed by the ec2 service with the associated policy : AmazonEC2ContainerServiceforEC2Role
* one role assumed by the batch service with the associated policy : AWSBatchServiceRole

So, the gist of my request is about knowing where I can access those roles, or an `IGrantable`, that I can give permissions to so a secret and a key can be utilized by the Batch service.

In other words, if you wanted to give `grant_read` for a secret and `grant_decrypt` for a key to your freshly created batch  jobs, how would you go about it ? See example above for Lambda.

Make sense ?

**Comment by madeline-k** [comment]
Hi @npvisual, sorry for the delay on this. Thanks for the additional explanation! I did a little bit more research into AWS Batch and can give you a better answer now. (Some of the below summary is for my own benefit.)

When a Job is submitted to AWS Batch, there are 4 IAM roles involved:

1. The ComputeEnvironment's InstanceRole, which can be provided in the `instance_role` parameter of the [`compute_resources`](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/ComputeResources.html#computeresources) parameter of ComputeEnvironment. If it is not by provided by the user, it is created automatically by the CDK [here](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-batch/lib/compute-environment.ts#L390) and has the AmazonEC2ContainerServiceforEC2Role policy attached. 
    - This role is described [here in the Batch docs](https://docs.aws.amazon.com/batch/latest/userguide/instance_IAM_role.html).
    - When the CDK creates this role, it has a construct id of 'Ecs-Instance-Role'
2. The ComputeEnvironment's ServiceRole, which can be provided in the `service_role` parameter documented [here](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/ComputeEnvironment.html). If it is not provided by the user in that parameter, then the CDK will generate it [here](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-batch/lib/compute-environment.ts#L407) and attach the AWSBatchServiceRole policy.
    - This role is described in [this doc](https://docs.aws.amazon.com/batch/latest/userguide/service_IAM_role.html) and is used by the Batch service to make calls to other AWS services on your behalf to manage the resources that you use with the service. You probably won't ever need to modify this one. 
    - When the CDK creates this role it has a construct id of 'Resource-Service-Instance-Role'.
3. The JobDefinition's ExecutionRole, which can be provided in the `execution_role` parameter of `JobDefinitionContainer` documented [here](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinitionContainer.html#aws_cdk.aws_batch.JobDefinitionContainer). 
    - If a role is not provided, the CDK does not create one for you. This role is required when your ComputeEnvironment uses Fargate, and the CDK will throw an error in that case and require you to provide one. If using EC2, I am not sure what exactly the behavior is for this one, but that is behavior determined by the AWS Batch service, not the CDK or Cloudformation.
    - This role is described in [this doc](https://docs.aws.amazon.com/batch/latest/userguide/execution-IAM-role.html). And is also equivalent to the the role described in [this page](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_execution_IAM_role.html) of the ECS developer guide. (AWS Batch jobs run in ECS, so the ExecutionRole defined in the JobDefinition will be used by ECS as the task execution role.) This role determines the capabilities of the ECS agent for example, allow it to pull images from ECR, and setup logging for your tasks.
4. The JobDefinitions's JobRole, can be provided in the `job_role` parameter of `JobDefinitionContainer` documented [here](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_batch/JobDefinitionContainer.html#aws_cdk.aws_batch.JobDefinitionContainer). 
    - This role is for specific capabilities within the task itself. It is used when your AWS Batch Job code actually runs. This is the role that needs to be modified if you want to grant your jobs permission to access other AWS resources.
    - Details of the role are documented [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) in the ECS docs. Like I mentioned above, Batch uses ECS to run jobs, so the JobRole that you provide to Batch will be in turn provided to ECS as the TaskRole.
    - If a role is not provided, the CDK does not create one for you. If no role is provided to CloudFormation and subsequently to AWS Batch when creating a JobDefinition, then your jobs will be launched with no Task Role and not have any permissions to perform any AWS API calls. 

In order to fulfill the request of modifying the `instance_role`, there are two ways.
1. The workaround that you did above using `find_child("Ecs-Instance-Role")`
2. Create your own role, and pass it into the `instance_role` parameter. Then you will already have a reference to the role and can modify the role as much as you want before or after passing it to the creation of the ComputeEnvironment. 

To modify the `job_role`, there is only one option: Create your own role, add the permissions that you want it to have. You can use `construct.grantReadXX(myRole)` APIs for this. Then pass it into the `job_role` parameter of your `JobDefinitionContainer`. You can continue to modify the role and give it more permissions after creation of the `JobDefinitionContainer` as well.

Hopefully this info unblocks you!

**Comment by madeline-k** [comment]
The initial issue description here is not a bug. What I was trying to say before is that any parameters that you pass into the constructor of a Construct are never modified. (Maybe I should say rarely modified instead. I don't know of any instances of this, but the CDK is large.) In this case the `JobDefinitionContainer` instance is simply a parameter to the `JobDefinition` construct. 

However, I could easily see this issue turned into a feature request. Here are a few options. 

1. Surface the InstanceRole (# 1 above) on the ComputeEnvironment class. This would mean that after creating your ComputeEnvironment, you could reference the role and add permissions to it. Python example:
```
computeEnvironment: batch.ComputeEnvironment = batch.ComputeEnvironment(...)
computeEnvironment.instance_role.add_managed_policy(iam.ManagedPolicy.from_aws_managed_policy_name('policyName'))
secret.grant_read(computeEnvironment.instance_role)
```

2. Surface the JobRole (# 4 above) on the JobDefinition class. This pretty similar to # 1. Python example:
```
jobDefinition: batch.JobDefinition = batch.JobDefinition(...)
jobDefinition.job_role.add_managed_policy(iam.ManagedPolicy.from_aws_managed_policy_name('policyName'))
secret.grant_read(jobDefinition.job_role)
```
3. Implement the `IGrantable` interface with ComputeEnvironment. This is a little bit more complicated feature request because how do you determine which of the two roles associated with a ComputeEnvironment should be the one used for `grant` requests? For example should `secret.grant_read(computeEnvironment)` grant read permissions to the InstanceRole or the ServiceRole?

4. Implement the `IGrantable` interface with JobDefinition. This one has the same problem of how do you determine which of the two roles associated with a JobDefinition should be the one used for `grant` requests? For example should `secret.grant_read(jobDefinition)` grant read permissions to the ExecutionRole or the JobRole?

**Comment by madeline-k** [comment]
I changed the title to reflect that I think we should implement feature requests # 1 and # 2 above. I don't see a way to resolve the ambiguity with feature requests # 3 and # 4. But definitely open to further discussion here!

**Comment by npvisual** [comment]
@madeline-k : any updates on timing for completion of this request ?

**Comment by comcalvi** [comment]
We've redesigned the batch constructs and have decided to not create a job role by default because the job role is very application-specific.

Making `JobDefinition` implement `IGrantable` suffers from one additional piece of ambiguity, in addition to what @madeline-k pointed out; multinode jobs have multiple container ranges, each with its own `jobRole` and `executionRole`. A user may want to grant permissions to one container range, but not another, within the same job definition, so these should *not* implement `IGrantable`.

The compute environment's `instanceRole` is now created by default (along with the profile needed to use it). If you use spot instances, the `spotFleetRole` is also created. The `serviceRole` and `spotFleetRole` are service-linked roles, and those only ever need very specific permissions. Using them as a target of `grant*()` does not make sense. (Besides, the `serviceRole` is created by Batch (not CFN) by default for managed compute environments, so granting to it is impossible; this is by design, as you should not be modifying the permissions of that role)

The `spotFleetRole` is a service linked role that we should create, if needed, and doesn't make sense to use for `IGrantable`. So that leaves only the `instanceRole`, which would make sense for `IGrantable`.

In short:

`JobDefinition`s are too ambiguous to use `IGrantable` with, because multi node jobs can contain multiple `jobRole`s. 
`ManagedComputeEnvironment`s could implement `IGrantable` with `instanceRole`.

Is it acceptable to you if I rename this issue to ask for managed compute environments to implement `IGrantable`?

**Comment by npvisual** [comment]
> @npvisual Is it acceptable if I rename this issue to ask for managed compute environments to implement IGrantable?

For sure (I see you already did). I have updated to CDK 2.81.0 and we're now using `FargateComputeEnvironment`. However the workaround we've used so far is failing : 

```
RuntimeError: Error: No child with id: 'Ecs-Instance-Role'
```

So this is really problematic since we have to upgrade the CDK and we don't seem to have a way to grant permissions to the batch compute environment (or instance) to read secrets or decrypt encryption keys.

**Comment by npvisual** [comment]
Any input on the question above ☝️ ?  @comcalvi / @madeline-k

**Comment by npvisual** [comment]
See also https://github.com/aws/aws-cdk/issues/25675
This is somewhat related to this issue as I had to add the following code to bypass the issue mentioned above : 

```python
        # Adding the ECS managed policy to gain access to ECS, ECR and Logs actions
        # This is the policy that would be found on the role that was previously
        # automatically created for the ECS instances prior to the breaking change.
        # See BE-2449.
        ecs_exec_role.add_managed_policy(
            iam.ManagedPolicy.from_aws_managed_policy_name(
                "service-role/AmazonEC2ContainerServiceforEC2Role"
            )
        )
```

**Comment by comcalvi** [comment]
Sorry didn't see this until now. Can you share the CDK definition of your CE and JobDefinition as well?

I've changed batch to create a role by default, because logging is enabled by default, so we should be able to add these grant methods now.

**Comment by npvisual** [comment]
I managed to make things work by adding the role to the `job_role` parameter of the `EcsFargateContainerDefinition`. I was previously just using the `execution_role`.

Ideally I'd like to pare it down, or better just use the one created for the `EcsFargateContainerDefinition` and not deal with fiddling with the permissions.

I saw the commit yesterday. Will try with the new bits as well.

See below for the info you requested and added the relevant parts as well with the key and secret so it gives a better understanding of what I am trying to achieve : 

```python
        # Compute environment
        batch_compute_environment = batch.FargateComputeEnvironment(
            self,
            f"{construct_id}-compute",
            vpc=vpc,
            replace_compute_environment=True,
            security_groups=[batch_security_group],
            terminate_on_update=False,
            vpc_subnets=ec2.SubnetSelection(subnets=subnets),
            compute_environment_name=f"{vanity}-compute-batch",
            enabled=True,
        )

....

        # ECS job role for the container (ecs-task).
        # We have to cheat our way out of :
        # https://github.com/aws/aws-cdk/issues/16020

        ecs_exec_role: iam.Role = iam.Role(
            self,
            f"pod-{vanity}-job-role",
            assumed_by=iam.ServicePrincipal("ecs-tasks.amazonaws.com"),
            description="ECS Container Role",
        )

        # Adding the ECS managed policy to gain access to ECS, ECR and Logs actions
        # This is the policy that would be found on the role that was previously
        # automatically created for the ECS instances prior to the breaking change.
        ecs_exec_role.add_managed_policy(
            iam.ManagedPolicy.from_aws_managed_policy_name(
                "service-role/AmazonEC2ContainerServiceforEC2Role"
            )
        )

        # Container definition (dropdb, createdb and adduser)
        batch_setup_container = batch.EcsFargateContainerDefinition(
            self,
            f"{vanity}-fargate-container-setup",
            assign_public_ip=False,
            cpu=2,
            memory=Size.gibibytes(4),
            image=ecs.ContainerImage.from_ecr_repository(repository=ecr_repo, tag=tripper_tag),
            command=[
                "sh",
                "-c",
                f"cmd_removed_proprietary_info",
            ],
            environment=environment_variables,
            execution_role=ecs_exec_role,
            job_role=ecs_exec_role,
        )
...
        # We give the batch compute resource role the proper permission
        # to read the DB cluster's secret (and therefore a `decrypt` permission
        # on the secret's key)
        database_secret.grant_read(ecs_exec_role)
        database_secret.encryption_key.grant_decrypt(ecs_exec_role)

        # We do the same for the shared database secret and its key...
        shared_database_key.grant_decrypt(ecs_exec_role)
        shared_database_secret.grant_read(ecs_exec_role)
```

**Comment by npvisual** [comment]
So ideally I'd like to be able to just use the following : 

```python
        # Container definition (dropdb, createdb and adduser)
        batch_setup_container = batch.EcsFargateContainerDefinition(
            self,
            f"{vanity}-fargate-container-setup",
            assign_public_ip=False,
            cpu=2,
            memory=Size.gibibytes(4),
            image=ecs.ContainerImage.from_ecr_repository(repository=ecr_repo, tag=tripper_tag),
            command=[
                "sh",
                "-c",
                f"cmd_removed_proprietary_info",
            ],
            environment=environment_variables,
        )
.... (note that we're not passing any exec or job role, just relying on the default being created)

        database_secret.grant_read(batch_setup_container.job_role.grant_principal)
        database_secret.encryption_key.grant_decrypt(batch_setup_container.job_role.grant_principal)
```

or something similar, where I can rely on the role automatically created by the `EcsFargateContainerDefinition` and just pass the `job_role.grant_principal`.

**Comment by comcalvi** [comment]
I see, this makes sense to me. I'll pick this up. 

> # This is the policy that would be found on the role that was previously
> # automatically created for the ECS instances prior to the breaking change.

This is a good callout, I think not using this managed policy is actually a source of the other permissions we've seen for the new constructs.

**Comment by npvisual** [comment]
@comcalvi just checking in to see if there was any progress made / plans available for when this would be implemented. Thanks !
